// contentful-batch-libs/dist/get/get-outdated-destination-content.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = getOutdatedDestinationContent;

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _npmlog = require('npmlog');

var _npmlog2 = _interopRequireDefault(_npmlog);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var BATCH_CHAR_LIMIT = 1990;
var BATCH_SIZE_LIMIT = 100;

/**
 * Gets content from a space which will have content copied to it, based on a
 * collection of existing content.
 *
 * Only the supplied entry/asset IDs will be retrieved. All contentTypes
 * and Locales will be retrieved.
 */
function getOutdatedDestinationContent(_ref) {
  var managementClient = _ref.managementClient,
      spaceId = _ref.spaceId,
      sourceResponse = _ref.sourceResponse,
      contentModelOnly = _ref.contentModelOnly,
      skipLocales = _ref.skipLocales,
      skipContentModel = _ref.skipContentModel;

  _npmlog2.default.info('Checking if destination space already has any content and retrieving it');

  return managementClient.getSpace(spaceId).then(function (space) {
    var result = {};
    var sourceContent = {
      contentTypes: sourceResponse.contentTypes || [],
      locales: sourceResponse.locales || [],
      entries: sourceResponse.entries || [],
      assets: sourceResponse.assets || []
    };

    if (!skipContentModel) {
      var contentTypeIds = sourceContent.contentTypes.map(function (e) {
        return e.sys.id;
      });
      result.contentTypes = batchedIdQuery(space, 'getContentTypes', contentTypeIds);

      if (!skipLocales) {
        var localeIds = sourceContent.locales.map(function (e) {
          return e.sys.id;
        });
        result.locales = batchedIdQuery(space, 'getLocales', localeIds);
      }
    }

    if (contentModelOnly) {
      return _bluebird2.default.props(result);
    }

    var entryIds = sourceContent.entries.map(function (e) {
      return e.sys.id;
    });
    var assetIds = sourceContent.assets.map(function (e) {
      return e.sys.id;
    });
    result.entries = batchedIdQuery(space, 'getEntries', entryIds);
    result.assets = batchedIdQuery(space, 'getAssets', assetIds);
    result.webhooks = [];

    return _bluebird2.default.props(result);
  }, function (err) {
    _npmlog2.default.error(`
The destination space was not found. This can happen for multiple reasons:
- If you haven't yet, you should create your space manually.
- If your destination space is in another organization, and your user from the source space does not have access to it, you'll need to specify separate sourceManagementToken and destinationManagementToken

Full error details below.
`);
    throw err;
  });
}

function batchedIdQuery(space, method, ids) {
  return _bluebird2.default.reduce(getIdBatches(ids), function (fullResponse, batch) {
    return space[method]({
      'sys.id[in]': batch,
      limit: batch.split(',').length
    }).then(function (response) {
      fullResponse = fullResponse.concat(response.items);
      return fullResponse;
    });
  }, []);
}

function getIdBatches(ids) {
  var batches = [];
  var currentBatch = '';
  var currentSize = 0;
  while (ids.length > 0) {
    var id = ids.splice(0, 1);
    currentBatch += id;
    currentSize = currentSize + 1;
    if (currentSize === BATCH_SIZE_LIMIT || currentBatch.length > BATCH_CHAR_LIMIT || ids.length === 0) {
      batches.push(currentBatch);
      currentBatch = '';
      currentSize = 0;
    } else {
      currentBatch += ',';
    }
  }
  return batches;
}
module.exports = exports['default'];
